### PR TITLE
Support for ManyToManyField in ClusterableModel.

### PR DESCRIPTION
**Pull request by theju**
Refs #31 which got closed unintentionally while rebasing the branch.

This is just an exploratory PR and may not match your coding conventions or styles. I will open a new one after I have satisfied all your concerns. Still supports django 1.8+ only.

### CODE REVIEW

**Comment by theju** [comment]
@gasman Refs torchbox/wagtail#231 and this PR, please let me know when if you need any clarifications or changes. More tests or docs or better code.

**Comment by gasman** [comment]
Hi @theju - sorry again for the delay in reviewing this, and thank you for taking this task on!

I see that the serialised representation of a model includes the related models as children. Using your test models as an example:

    {
        "pk": null,
        "authors": [{"pk": 1, "name": "Bill"}, {"pk": 2, "name": "Ted"}],
        "categories": [],
        "title": "hello world"
    }

Here it's possible for the author records to be shared between articles. This doesn't really fit with the existing concepts of modelcluster, or the desired behaviour in Wagtail. (In Wagtail we store the JSON serialisation for each page revision, so if an author is linked to two pages and the name is changed on one of them, the author record in the database will get overwritten as one page or the other is published.)

Modelcluster is designed on the principle that the related models 'belong' to the parent ClusterableModel, and the whole cluster behaves as a single standalone unit, with no shared data. On that basis, I think the best approach is for "authors" to be a list of IDs (or 'through' models, if the M2MField defines a `through` parameter):

    {
        "pk": null,
        "authors": [1, 2],
        "categories": [],
        "title": "hello world"
    }

or (with a field definition like `authors = M2MField('Author', through=ArticleAuthorship, related_name="articles_by_author")`):

    {
        "pk": null,
        "authors": [{"pk": 101, "author_id": 1, role: "original text"}, {"pk": 102, "author_id": 2, role: "translation"}],
        "categories": [],
        "title": "hello world"
    }

This does mean there's no way to attach *unsaved* related objects via an M2M relation, but I don't think that's necessary. In Wagtail, the use-case for M2M relations would be things like image galleries or 'related pages', where the items on the other side of the relation are images/pages/snippets that already exist in the database, selected with a chooser. (Modelcluster shouldn't be designed purely around Wagtail's requirements, of course... but I can't really see a situation where you'd be working with unsaved 'child' objects, *and* want them to be managed as an M2M relation.)

I believe this would simplify the implementation, too - I *think* it would let us eliminate `M2MRelatedObjectsDescriptor`, because the deferring_many_related_manager behaviour would only be needed on one side of the relation. Effectively, the relation would be 'owned' by the model that defines it - I would suggest naming the field `ParentalManyToManyField` for this reason.

    class Article(ClusterableModel):
        title = models.CharField(max_length=255)
        authors = ParentalManyToManyField('Author', related_name="articles_by_author")
        categories = ParentalManyToManyField('Category', related_name="articles_by_category")

`article.authors = [tom, dick, harry]` would have the deferring_many_related_manager behaviour, but `author.articles_by_author = [article1, article2]` would behave as a plain M2M relation - which I think would be fine, because you wouldn't normally be modifying the relation from that side. Does that sound reasonable?

**Comment by theju** [comment]
@gasman Thanks for the comments.

I was surprised to learn that the PageRevision model doesn't store the completely serialized data from ParentalKey fields. It would've served a purpose as an audit log. But besides that point, I think your comment clarifies a lot for me.

I will work up a new PR with the changes you've suggested.

**Comment by gasman** [comment]
For ParentalKey fields, we do store the full serialized data - those related objects are not shared with any other pages, so it's safe to consider them as part of the overall page record (without any danger of overwriting / duplicating data). I propose that M2M fields should behave differently to ParentalKeys, and just store the IDs - since they conceptually refer to data that exists outside of the record.

**Comment by theju** [comment]
If we consider (and I agree with your assumption) that if M2M fields are outside of the record, shouldn't we be storing more than just the primary keys for the ManyToMany fields in the serialized data as you've suggested in the authors snippet?

**Comment by gasman** [comment]
I'm not sure what you mean - if the M2M fields are outside of the record, then surely that's a reason to reference them as IDs, rather than storing data from those objects in the serialized data.

(You're talking about the first example with `"authors": [1, 2]`, right...?)

**Comment by theju** [comment]
I have started work based on your comments at https://github.com/theju/django-modelcluster/tree/parentalmanytomany-field

There is one problem while generating the model from the serialized data it expects a dictionary with a 'pk' key [1] but we would only be storing a list of primary keys. What do you suggest?

[1] https://github.com/torchbox/django-modelcluster/blob/master/modelcluster/models.py#L60

**Comment by gasman** [comment]
I think the best plan would be to special-case ParentalManyToManyField fields inside ClusterableModel.from_serializable_data, so that we perform a simple ID lookup on the list instead of running it through model_from_serializable_data. Something like:


            related_model = get_related_model(rel)

            if rel.many_to_many:
                children = related_model.objects.filter(pk__in=child_data_list)
            else:
                if hasattr(related_model, 'from_serializable_data'):
                    children = [
                        related_model.from_serializable_data(child_data, check_fks=check_fks, strict_fks=True)
                        for child_data in child_data_list
                    ]
                else:
                    children = [
                        model_from_serializable_data(related_model, child_data, check_fks=check_fks, strict_fks=True)
                        for child_data in child_data_list
                    ]

**Comment by theju** [comment]
I have made the change in the https://github.com/theju/django-modelcluster/tree/parentalmanytomany-field with the only change of using `related_model._default_manager.filter(pk__in=child_data_list)` as opposed to `related_model.objects.filter(pk__in=child_data_list)`.

All tests pass at this moment.

**Comment by theju** [comment]
@gasman Now that django 1.7 has been deprecated in Wagtail. Do you think this can be merged before wagtail v1.4 is out?

**Comment by gasman** [comment]
@theju That's the plan, yep.

**Comment by theju** [comment]
@gasman I am starting this work on this PR afresh as a lot has changed between django 1.8 and 1.9 and making it work across the two versions quite complicated.

**Comment by theju** [comment]
@gasman I have the tests passing on django 1.8 and 1.9 now. The code is on the [parentalmanytomany-field branch](https://github.com/theju/django-modelcluster/tree/parentalmanytomany-field). Please have a look and review the code.

**Comment by auvipy** [comment]
@theju all the checks are failing. could you fix them?

**Comment by gasman** [comment]
@auvipy The Travis checks here are massively outdated - in particular, we've now dropped Django 1.7 support.

**Comment by gasman** [comment]
parentalmanytomany-field branch rebased on top of current master at: https://github.com/gasman/django-modelcluster/tree/m2m

**Comment by theju** [comment]
@gasman Saw the rebased branch. What's the next step? Do you require any changes from me?

**Comment by gasman** [comment]
@theju I'm reviewing it now. I just noticed that test_deserialize is failing on postgres, though: https://travis-ci.org/gasman/django-modelcluster/builds/124987013

**Comment by gasman** [comment]
test_deserialize issue on Postgres fixed in https://github.com/gasman/django-modelcluster/commit/1eaa5b9eb0f1b7aefa2d3932742434e3b455f233

**Comment by theju** [comment]
@gasman Thanks for reviewing the patch. I am on standby if you require anything from me.

File: modelcluster/fields.py [hunk 1]
```diff
@@ -270,3 +271,278 @@ def check(self, **kwargs):
...
+        super(M2MField, self).contribute_to_related_class(cls, related)
+
+    def get_accessor_name(self):
+        return self.name
+
+    def _get_val_from_obj(self, obj):
+        mgr = super(M2MField, self)._get_val_from_obj(obj)
+        return mgr.get_object_list()
+
+    def get_searchable_content(self, value):
```
Comment (gasman) [inline modelcluster/fields.py]: I think this should be left out. `get_searchable_content` is only used by Wagtail's search mechanism, and we should keep Wagtail-specific functionality out of modelcluster.

In any case, indexing a list of PKs probably isn't going to be very useful for searching (and it's hard to know whether indexing the actual content of the related model would be a good thing or not... see the discussion at https://github.com/torchbox/wagtail/issues/2500#issuecomment-212158853). Hopefully [`index.RelatedFields`](http://docs.wagtail.io/en/v1.4.3/topics/search/indexing.html#index-relatedfields) already handles M2M relations - or can be updated to do so - which would mean we don't need any special treatment of search indexing here.

**Comment by gasman** [comment]
Hi @theju,

I've added a test for serialising a ParentalManyToManyField here:

https://github.com/gasman/django-modelcluster/commit/7c5b3fdaf842cfe5a5b772e0438399bf9455a5e4

This is currently failing on two counts:

* it can't retrieve the `categories` relation unless I explicitly initialise it with an empty list (`categories=[]`) - presumably if I leave it uninitialised, it's falling back to the real queryset and failing because the article is unsaved
* the authors are being serialised to the complete objects, rather than just their PKs.

**Comment by theju** [comment]
With respect to the categories raising an error when uninitialized, we are running into https://github.com/django/django/blob/master/django/db/models/fields/related_descriptors.py#L756 which doesn't seem to be the case for `RelatedManager.` What is the best way to handle this? Do we go over all the fields and set uninitialized fields to `[]`?

For the authors being serialized to complete instances, should we special case `get_serializable_data_for_fields` or the `serializable_data` method?

**Comment by theju** [comment]
@gasman I have fixed the failing m2m serialisation test in https://github.com/theju/django-modelcluster/commit/83a1ee12848b4cb95d203931a94dc30b1040a78e and with a slight correction in https://github.com/theju/django-modelcluster/commit/787303598078625cebb4f2167993bb577548836e

Now all tests pass for django 1.8 and 1.9. Please have a look.

**Comment by alexgleason** [comment]
Any updates? Thanks all for your work

**Comment by theju** [comment]
@alexgleason I believe I have made the changes suggested by @gasman and waiting for a review.

**Comment by auvipy** [comment]
@theju you need to rebase you PR on master

**Comment by theju** [comment]
@auvipy I have pushed the changes on top of master but some unrelated tests are failing which I am not seeing on my local while running `python runtests.py`

**Comment by gasman** [comment]
The test failure was due to a breakage in newer versions of django-taggit, and should be fixed by #59.

**Comment by gasman** [comment]
Hi @theju - unfortunately Wagtail's test suite is failing when running against this version of modelcluster. It looks like this is due to the change to `get_all_child_relations` https://github.com/torchbox/django-modelcluster/pull/36/files#diff-83a16c5a484cd2a24a226074993a2988R136 - you have it returning all M2M relations, not just ones declared with ParentalManyToManyField. (In Wagtail's case, this includes ClusterTaggableManager fields, which set the `many_to_many` flag to pretend to be M2M relations.)

With this fix in place we're probably still going to encounter problems with existing code that assumes that `get_all_child_relations` will only ever return `ManyToOneRel` descriptors. For example, https://github.com/torchbox/wagtail/blob/5ce99e85c413fb13743d337f6faa5addcec61b4d/wagtail/api/v2/serializers.py#L330 expects to find a `field` attribute, which `ManyToManyField` descriptors don't have. I suspect we'll need to introduce a separate `get_all_child_m2m_relations` utility function. (If necessary, we can break backwards compatibility and release this as modelcluster 3.0 - I'd prefer to avoid that, though.)

File: tests/tests/test_cluster.py [hunk 2]
```diff
@@ -278,3 +279,59 @@ class Meta:
...
+    def test_parentalm2mfield(self):
+        article = Article(title="Test Title")
+        author_1 = Author(name="Author 1")
+        author_2 = Author(name="Author 2")
+        article.authors = [author_1, author_2]
+        category_1 = Category(name="Category 1")
+        category_2 = Category(name="Category 2")
+        article.categories = [category_1, category_2]
+        self.assertEqual(['Author 1', 'Author 2'],
+                         [author.name for author in article.authors.all()])
```
Comment (gasman) [inline tests/tests/test_cluster.py]: We can't guarantee the ordering here (well, we probably _can_ guarantee it for our fake querysets, but not live ones coming from the database) - there should probably be an `order_by('name')` here, or else we should compare them as sets.

File: modelcluster/models.py [hunk 3]
```diff
@@ -186,7 +190,10 @@ def save(self, **kwargs):
         super(ClusterableModel, self).save(update_fields=real_update_fields, **kwargs)
 
         for relation in relations_to_commit:
-            getattr(self, relation).commit()
+            try:
+                getattr(self, relation).commit()
+            except AttributeError:
+                pass
```
Comment (gasman) [inline modelcluster/models.py]: If you're going to catch and discard exceptions, please add a comment and relevant test coverage to show why this is a valid thing to do, and isn't simply hiding a genuine error.

In this case, I don't think it is valid: it's just masking the bug described in https://github.com/torchbox/django-modelcluster/pull/36#issuecomment-237304134 (the lack of `commit` method is an indication that `get_all_child_relations` is incorrectly returning things that aren't `Deferring*RelatedManager`s).

File: modelcluster/forms.py [hunk 4]
```diff
@@ -177,12 +177,15 @@ def __new__(cls, name, bases, attrs):
...
-                }
+                try:
+                    kwargs = {
+                        'extra': cls.extra_form_count,
+                        'formfield_callback': formfield_callback,
+                        'fk_name': rel.field.name,
+                        'widgets': widgets
+                    }
+                except AttributeError:
+                    continue
```
Comment (gasman) [inline modelcluster/forms.py]: Same issue as https://github.com/torchbox/django-modelcluster/pull/36/files#r73386914. If we need special-case handling for M2M fields here, there should be a test for that. (As it is, if I remove this try/except, the only test failures are ones relating to `ClusterTaggableManager`.)

**Comment by theju** [comment]
@gasman I have made the changes you have suggested. Please have a look at the changes.

**Comment by havard024** [comment]
Any progress on this pull request?

I'm in need of this exact fix for a project and I'm willing to assist if you need any help.

**Comment by gasman** [comment]
@havard024 It's waiting for me to review it. In the meantime, you're more than welcome to review the code yourself, try it out against your project and let us know any feedback.

**Comment by gasman** [comment]
Thanks again @theju, and sorry for the delay in reviewing.

There were still some issues with the fix for https://github.com/torchbox/django-modelcluster/pull/36#issuecomment-237304134 - the result of `get_all_child_relations` needs to skip all non-parental M2M fields, not just ClusterTaggableManager, and in any case it's better to preserve backwards compatibility by adding a new `get_all_child_m2m_relations` method rather than extending the existing one. I've now made this change, along with various other updates, in #62.

**Comment by theju** [comment]
@gasman Thanks. Yay! No need to maintain a fork for this any more.

**Comment by auvipy** [comment]
@gasman  can we expect a new release?

**Comment by gasman** [comment]
@auvipy Yep - I've now built version 3.0a1 and pushed to PyPI. I'll release 3.0 final some time in the next month, once it's had some more testing with Wagtail (and before the release of Wagtail 1.9).

**Comment by auvipy** [comment]
:dancing_men:

**Comment by alexgleason** [comment]
This is awesome news!

**Pull request closed without merging**